api/controller.py

```diff
@@ -1478,3 +1499,248 @@ def notify(self, loan_id):
+        if isinstance(client, ProblemDetail):
+            return client
+        loan = get_one(self._db, Loan, id=loan_id, integration_client=client)
+        if not loan or not loan.license_pool.collection == collection:
+            return NO_ACTIVE_LOAN
+
+        try:
+            self.shared_collection.revoke_loan(collection, client, loan)
+        except AuthorizationFailedException, e:
+            return INVALID_CREDENTIALS.detailed(str(e))
+        except NotCheckedOut, e:
+            return NO_ACTIVE_LOAN.detailed(str(e))
+        except CannotReturn, e:
+            return COULD_NOT_MIRROR_TO_REMOTE.detailed(str(e))
+        return Response(_("Success"), 200)
+
+    def fulfill(self, collection_name, loan_id, mechanism_id, do_get=HTTP.get_with_timeout):
+        collection = self.load_collection(collection_name)
+        if isinstance(collection, ProblemDetail):
+            return collection
+        client = self.authenticated_client_from_request()
+        if isinstance(client, ProblemDetail):
+            return client
+        loan = get_one(self._db, Loan, id=loan_id)
+        if not loan or not loan.license_pool.collection == collection:
+            return NO_ACTIVE_LOAN
+
+        mechanism = None
+        if mechanism_id:
+            mechanism = self.load_licensepooldelivery(
+                loan.license_pool, mechanism_id
+            )
+            if isinstance(mechanism, ProblemDetail):
+                return mechanism
+            
+        if not mechanism:
+            # See if the loan already has a mechanism set. We can use that.
+            if loan and loan.fulfillment:
+                mechanism = loan.fulfillment
+            else:
+                return BAD_DELIVERY_MECHANISM.detailed(
+                    _("You must specify a delivery mechanism to fulfill this loan.")
+                )
+
+        try:
+            fulfillment = self.shared_collection.fulfill(collection, client, loan, mechanism)
+        except AuthorizationFailedException, e:
+            return INVALID_CREDENTIALS.detailed(str(e))
+        except CannotFulfill, e:
+            return CANNOT_FULFILL.detailed(str(e))
+        except RemoteIntegrationException, e:
+            return e.as_problem_detail_document(debug=False)
+        headers = dict()
+        content = fulfillment.content
+        if fulfillment.content_link:
+            # If we have a link to the content on a remote server, web clients may not
+            # be able to access it if the remote server does not support CORS requests.
+            # We need to fetch the content and return it instead of redirecting to it.
+            try:
+                response = do_get(fulfillment.content_link)
+                status_code = response.status_code
+                headers = dict(response.headers)
+                content = response.content
+            except RemoteIntegrationException, e:
+                return e.as_problem_detail_document(debug=False)
```

Comment:
This is fine for now, but since we have no control over the size of the thing at the other end of the link, I would eventually like to get to the point where the client can signal whether or not it needs us to do this.

Reply:
Sure, and the same applies to regular fulfills.